### ISSUE TITLE
Tall character cut-off (and weird virtual character position?)

### ISSUE THREAD

**Issue by JonasUJ** [Kethku/neovide#533]
Left part of image is Neovide (I get the same behavior in Alacritty on both Windows and Ubuntu)
Right part of image is Windows Terminal (I get the same behavior in gnome-terminal on Ubuntu)

![image](https://user-images.githubusercontent.com/21989288/114305017-8af62c80-9ad6-11eb-905e-5941deff4e0e.png)

The things I've marked in orange are tall characters being cut off at the top and bottom. The characters in question are `▎` and `▏`.

Another thing to note is that Windows Terminal renderes the characters in the center of the column (look at the cursor), even though they are positioned to the left.

The font I'm using is CaskaydiaCove NF (Default Windows Terminal font Cascadia Code patched with NerdFonts)

The thing marked with green is the previously uncentered character moving to the center when... something happens. I'm using the Lua branch of "lukas-reineke/indent-blankline.nvim" for the indentlines. I think it uses virtual text to show them. If I move the cursor up one line this happens (duplicate characters)

![image](https://user-images.githubusercontent.com/21989288/114305228-749ca080-9ad7-11eb-90d2-4c8bf12d0fe0.png)
 
Since I can reproduce both of these things in multiple terminals, I don't know which has the right behavior, but duplicate characters does seem like a bug in Neovide.

**Comment by Kethku** [comment]
Ok this is very related to the some font shaping issues I've been running into. As you've noted, its two problems

1. positioning of characters vertically may need to be adjusted. Its also interesting that the line character in windows terminal is thinner but matches closer to the pipe character. Almost feels like different glyphs are being selected by the shaper
2. the shaping is not treating character positions as monospace. This something I've written a bit about here: https://github.com/Kethku/neovide/issues/491#issuecomment-815330472. I think the solution is to use the harfbuzz shaper only for glyph selection, and handle spacing of the characters ourselves because we know neovim only expects monospace. This will look strange with variable width characters, but should fix issues like the one you've highlighted.

In the meantime, I will download the font you linked and see if I can repro the problems myself to dig in further.

**Comment by JonasUJ** [comment]
This is the exact link i downloaded it with https://github.com/ryanoasis/nerd-fonts/releases/download/v2.1.0/CascadiaCode.zip

**Comment by ErikMayrhofer** [comment]
I believe I may have stumbled upon a related problem:

I tried to set my guifont to `JetBrains Mono Nerd Font` through
```vim
" init.vim
set guifont=JetBrains\ Mono\ Nerd\ Font
```

After opening Neovide I noticed that some character seem to be shifted upwards and cut off. 
![image](https://user-images.githubusercontent.com/43534802/119237769-dc1a2700-bb3e-11eb-808d-17d64a091bde.png)


What I also noticed is that if I used the non-nerd version of the font
```vim
" init.vim
set guifont=JetBrains\ Mono
```
that issue does not come up:
![image](https://user-images.githubusercontent.com/43534802/119237799-01a73080-bb3f-11eb-9e9b-26fb6be2ec8e.png)
(This also works when switching away from the nerd-font at runtime with `:set guifont=JetBrains\ Mono`)

The peculiar thing is, that if I afterwards switch back to the nerd-font at runtime using `:set guifont=JetBrains\ Mono\ Nerd\ Font`, the issue seems to be magically resolved.
![image](https://user-images.githubusercontent.com/43534802/119237839-4337db80-bb3f-11eb-8eef-7fda7eb429c1.png)
But notice how now the GitHub-logo at the fourth line is cut off horizontally.

PS: I hope this is actually related and posting this here is the right place. If you think this is a separate issue I will happily open another one. I will probably investigate this further and post here if I come upon any things that may be of interest.
